### ISSUE TITLE
Triton Kernel 撰写

### ISSUE THREAD

**Issue by lingebeng** [lingebeng/op_dev#1]
撰写常见的 Triton Kernel，快速补齐对 Triton 对理解，以便更好的撰写 Pallas Kernel.

**Comment by lingebeng** [comment]
## Day 01 Softmax Kernel

```python
import torch
import triton
import triton.language as tl


@triton.jit
def softmax_kernel(
    inp_ptr, out_ptr, stride_inp_row, stride_out_row, N, BLOCK_SIZE: tl.constexpr
):
    row_idx = tl.program_id(0)

    inp_row_ptr = inp_ptr + row_idx * stride_inp_row
    out_row_ptr = out_ptr + row_idx * stride_out_row

    offsets = tl.arange(0, BLOCK_SIZE)

    inp_ptrs = inp_row_ptr + offsets
    out_ptrs = out_row_ptr + offsets

    mask = offsets < N
    row = tl.load(inp_ptrs, mask=mask, other=-float("inf"))

    row_max = tl.max(row, axis=0)

    row = tl.exp(row - row_max)

    row_sum = tl.sum(row,axis=0)

    out = row / row_sum

    tl.store(out_ptrs, out, mask=mask)


def softmax(inp):
    B, L = inp.shape

    out = torch.empty_like(inp)

    BLOCK_SIZE = triton.next_power_of_2(L)

    grid = (B,)

    softmax_kernel[grid](
        inp, out, inp.stride(0), out.stride(0), N=L, BLOCK_SIZE=BLOCK_SIZE
    )

    return out


if __name__ == "__main__":
    torch.manual_seed(0)
    inp = torch.randn(1024, 2048, device="cuda")
    out = softmax(inp)
    out_ref = torch.nn.functional.softmax(inp, dim=-1)

    max_diff = torch.max(torch.abs(out - out_ref))

    print(max_diff.item())

```

**Comment by lingebeng** [comment]
## Day 02 RMSNorm 

```python
@triton.jit
def rmsnorm_kernel(
    inp_ptr,
    out_ptr,
    w_ptr,
    stride_inp_0,
    stride_inp_1,
    stride_out_0,
    stride_out_1,
    D,
    eps,
    block_size: tl.constexpr,
):
    row_idx = tl.program_id(0)

    inp_row_ptr = inp_ptr + row_idx * stride_inp_0
    out_row_ptr = out_ptr + row_idx * stride_out_0

    s = 0.0

    for offset in range(0, D, block_size):
        cols = offset + tl.arange(0, block_size)
        mask = cols < D

        inp_chunk = tl.load(inp_row_ptr + cols * stride_inp_1, mask=mask, other=0.0).to(
            tl.float32
        )
        s += tl.sum(inp_chunk * inp_chunk, axis=0)

    rsqrt = 1.0 / tl.sqrt((s / D) + eps)

    for offset in range(0, D, block_size):
        cols = offset + tl.arange(0, block_size)
        mask = cols < D

        inp_chunk = tl.load(inp_row_ptr + cols * stride_inp_1, mask=mask, other=0.0)
        inp_chunk_fp32 = inp_chunk.to(tl.float32)

        w_chunk = tl.load(w_ptr + cols, mask=mask, other=0.0)

        out_chunk_fp32 = inp_chunk_fp32 * rsqrt
        out_chunk = out_chunk_fp32.to(w_chunk.dtype) * w_chunk

        tl.store(out_row_ptr + cols * stride_out_1, out_chunk, mask=mask)
```

Gemini 优化
```python
# 🚀 性能优化：在寄存器中分配向量
s_vec = tl.zeros([block_size], dtype=tl.float32)

for offset in range(0, D, block_size):
    cols = offset + tl.arange(0, block_size)
    mask = cols < D
    inp_chunk = tl.load(inp_row_ptr + cols * stride_inp_1, mask=mask, other=0.0).to(tl.float32)
    
    # 独立运算：每个线程只管算自己寄存器里的数字，绝对不和其他线程说话
    s_vec += inp_chunk * inp_chunk

# 循环彻底结束后，做唯一一次跨线程规约
s = tl.sum(s_vec, axis=0)
```

```bash
reduction-compare:
        T  PyTorch (GB/s)  Triton (Sum inside) (GB/s)  Triton (Vector Opt) (GB/s)
0   512.0      233.509850                 1261.824334                 1497.965768
1  1024.0      238.583836                 1535.250321                 1624.439904
2  1536.0      247.344542                 1717.100456                 1738.931970
3  2048.0      254.154023                 1774.240216                 1784.051062
4  2560.0      257.926896                 1707.778450                 1762.905142
5  3072.0      260.991293                 1801.676952                 1836.921458
6  3584.0      262.951631                 1857.295629                 1884.959385
7  4096.0      264.378837                 1892.736368                 1899.809318
```
